e2e: serial enhancements

shajmakh opened the pull request:
e2e serial suite enhancements. please check the commits

shajmakh committed:
e2e: serial: ensure MCPs and nodes are healthy

At the setup of the test, check that MCPs and nodes are healthy enough
for the test to run by verifying that the MCPs are all `Updated` and non of the nodes is
`unschedulable`. We don't want to fail on any error so keep any
unexpected results at warning level.

Signed-off-by: Shereen Haj <shajmakh@redhat.com>

shajmakh committed:
e2e:serial: fix baseload addition to pod resources

The old method used to apply the baseload on the same `paddingRes` over
and over again. This is not causing issues as long as there are no more
than one node to be padded, which is almost always the case on the MNO
used in the weekly CI.
However this is still incorrect, so this commit fixes it so that the
padding resources with the baseload are cleanly computed for each node.

Signed-off-by: Shereen Haj <shajmakh@redhat.com>

shajmakh committed:
e2e:serial: apply the baseload on the first numa zone

with the old version, the baseload considered in the padding resources
of every numazone. that is incorrect because it can create issues on
tests that expect their testing pods to be scheduled on one node only
while other nodes are unsuitable due to insufficient resources.
This update makes sure that only the forst NUMA zone s padded leaving
extra space for the baseload.

Signed-off-by: Shereen Haj <shajmakh@redhat.com>

ffromani commented on test/e2e/serial/tests/workload_placement_tmpol.go:
```diff
@@ -154,9 +154,14 @@ var _ = Describe("[serial][disruptive][scheduler] numaresources workload placeme
+					}
 					podName := fmt.Sprintf("padding-%d-%d", nIdx, zIdx)
-					padPod, err := makePaddingPod(fxt.Namespace.Name, podName, zone, paddingResWithBaseload)
+					padPod, err := makePaddingPod(fxt.Namespace.Name, podName, zone, zonePaddingRes)
```
there is no correct way to account for baseload per-NUMA because the baseload is composed by burstable and best-effort pods (most notably burstable) which can move freely around NUMA zones. Accounting against any NUMA zone is never fully correct.

This means that every possible solution is unavoidably partial and has pros and cons.

shajmakh committed:
e2e:serial: pull daemonset object in eventually to ensure updates

Sometimes it takes longer time to have the daemonset updates its pods
amounts. The problem with the old code is that in an Eventually it waits
for the daemonset pods number to get updated to the expected number
while using an old daemonset object. To fix this, just pull the
daemonset object in every iteration to ensure the code in the eventaully
is using the most updated data.

Signed-off-by: Shereen Haj <shajmakh@redhat.com>

ffromani commented on test/e2e/serial/tests/tolerations.go:
```diff
@@ -469,10 +469,11 @@ var _ = Describe("[serial][disruptive][rtetols] numaresources RTE tolerations su
 				// and the pod will take nonzero time to terminate, so we need a Eventually block.
+				By(fmt.Sprintf("ensuring the RTE DS is running with less pods because taints (expected pods=%v)", len(workers)-1))
 				Eventually(func(g Gomega) {
+					Expect(fxt.Client.Get(ctx, dsKey.AsKey(), updatedDs)).To(Succeed())
 					pods, err = podlist.With(fxt.Client).ByDaemonset(ctx, *updatedDs)
 					Expect(err).ToNot(HaveOccurred(), "failed to get the daemonset pods %s: %v", dsKey.String(), err)
```
let's use g.Expect() here like in the rest of the `Eventually`

ffromani commented on test/e2e/serial/tests/tolerations.go:
```diff
@@ -469,10 +469,11 @@ var _ = Describe("[serial][disruptive][rtetols] numaresources RTE tolerations su
 
 				// NoExecute promises the pod will be evicted "immediately" but the system will still need nonzero time to notice
 				// and the pod will take nonzero time to terminate, so we need a Eventually block.
+				By(fmt.Sprintf("ensuring the RTE DS is running with less pods because taints (expected pods=%v)", len(workers)-1))
```
gonna address in a sweeping AI-assisted PR

ffromani commented:
Continues in #3604 

we will address leftovers in future separate PRs
